Enforce global type declaration for `defaultOptions` that influence results (#13132)
apollographql/apollo-client: docs/source/data/typescript.mdx

Reviewer: `DeclaredDefaultOptions` supports a union for those values right? Would it make sense to add a section showing the union or do we want to avoid that so that we don't encourage that behavior?

});
```

## Declaring default options for type safety

Certain default options can affect the return types of methods like `client.query` and hooks like `useQuery` or `useSuspenseQuery`.
When you set `errorPolicy` or `returnPartialData` in your `ApolloClient` `defaultOptions`, declare them at the type level so TypeScript can accurately type your results.

### Why to declare default options

> The following example demonstrates this behavior:

Author: ```suggestion
Defining default options can alter the return type of hooks and client methods.
The following example demonstrates this behavior and why it is important to also declare these default options at the type level:
```

    namespace DeclareDefaultOptions {
      interface WatchQuery {
        errorPolicy: "all";
      }
      interface Query {
        errorPolicy: "all";
      }
      interface Mutate {
        errorPolicy: "all";
      }
    }
  }
}
```

Once declared, results will be typed correctly based on your defaults:

```ts
// With errorPolicy: "all" declared as default
const result = await client.query({ query: MY_QUERY });
result.data; // Type: TData | undefined
result.error; // Type: ErrorLike | undefined

// Override the default to get different types
const result2 = await client.query({ query: MY_QUERY, errorPolicy: "none" });
result2.data; // Type: TData
result2.error; // Type: never
```

Options that can be declared:

- **`WatchQuery`**: `errorPolicy`, `returnPartialData`
- **`Query`**: `errorPolicy`
- **`Mutate`**: `errorPolicy`

Declaring these options ensures you provide values for them in your `defaultOptions`.
If you don't declare these options, TypeScript prevents you from using these specific fields (for example, `errorPolicy` or `returnPartialData`) in the `ApolloClient` constructor's `defaultOptions`, but you can still provide other default options.

> If you can't declare specific global defaults (for example, if you have multiple Apollo Client instances with different default options), you can opt out by declaring the options as `unknown`:

Reviewer 2: ```suggestion
If you can't declare specific global defaults (for example, if you have multiple Apollo Client instances with different default options), you can opt out by declaring options as `unknown`:
```

```ts title="apollo.d.ts"
import type {} from "@apollo/client";

declare module "@apollo/client" {
  namespace ApolloClient {
    namespace DeclareDefaultOptions {
      interface WatchQuery {
        errorPolicy?: unknown;
        returnPartialData?: unknown;
      }
      interface Query {
        errorPolicy?: unknown;
      }
      interface Mutate {
        errorPolicy?: unknown;
      }
    }
  }
}
```

This will allow you to call the `ApolloClient` constructor with any arbitrary `defaultOptions`.

> In that case, all methods and hooks will assume that these `defaultOptions` are set to the defaults, so you might get incorrect return values.

Reviewer 2: ```suggestion
In that case, all methods and hooks assume that those `defaultOptions` are set to the defaults, so you might get incorrect return values.
```

});
```

> ## Declaring default options for type safety

Reviewer: ```suggestion
<MinVersion version="4.2.0">

## Declaring default options for type safety

</MinVersion>
```

### Why to declare default options

Defining default options can alter the return type of hooks and client methods.
The following example demonstrates this behavior and why it is important to also declare these default options at the type level:

```tsx
const client = new ApolloClient({
  // ...
  defaultOptions: {
    query: {
      errorPolicy: "all",
    },
  },
});

const result = await client.query({ query: MY_QUERY });
```

With `errorPolicy: "all"`, when a GraphQL error occurs, both `data` and `error` can be present simultaneously. More importantly, `data` can be `undefined` if the error prevents any data from being returned. Without declaring this default option at the type level, TypeScript would incorrectly type `data` as non-nullable, leading to potential runtime errors.

### How to declare default options

Declare your default options using TypeScript's [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) with the `ApolloClient.DeclareDefaultOptions` namespace. Create a TypeScript declaration file in your project:

```ts title="apollo.d.ts"

> import type {} from "@apollo/client";

Reviewer: Is the `type {}` necessary here? We recommend plain `import "@apollo/client"` in our [data masking docs](https://www.apollographql.com/docs/react/data/fragments#setting-data-masking-types).

I'd love to keep consistency. If the `type {}` is needed, can you update the data masking docs to reflect as well?

declare module "@apollo/client" {
  namespace ApolloClient {
    namespace DeclareDefaultOptions {
      interface WatchQuery {
        errorPolicy: "all";
      }
      interface Query {
        errorPolicy: "all";
      }
      interface Mutate {
        errorPolicy: "all";
      }
    }
  }
}
```

> Once declared, results will be typed correctly based on your defaults:

Reviewer: Does declaring these types require that you provide the default options to the `ApolloClient` constructor? If so, you might want to add a small section before this sentence that says something along the lines of "this will result in a TypeScript error in the `ApolloClient` constructor. Add these options [code block]."

Since this is a new section called "How to declare default options", I think it best to assume its a clean break from the previous section and shouldn't assume `defaultOptions` are already provided.

Might also be nice to show a sample of what the TypeScript error looks like in the ApolloClient constructor (replace with the real error message)

```ts
// ❌ Property 'defaultOptions' is missing in type `ApolloClient.Options`
const client = new ApolloClient({

})
```

```ts
// With errorPolicy: "all" declared as default
const result = await client.query({ query: MY_QUERY });
result.data; // Type: TData | undefined
result.error; // Type: ErrorLike | undefined

// Override the default to get different types
const result2 = await client.query({ query: MY_QUERY, errorPolicy: "none" });
result2.data; // Type: TData
result2.error; // Type: never
```

Options that can be declared:

- **`WatchQuery`**: `errorPolicy`, `returnPartialData`
- **`Query`**: `errorPolicy`

> - **`Mutate`**: `errorPolicy`

Reviewer: Is this specific to the global type declarations or `defaultOptions` itself? You might want to clarify that here to avoid confusion:

```suggestion
The following default options can be declared in `ApolloClient.DeclareDefaultOptions`:

- **`WatchQuery`**: `errorPolicy`, `returnPartialData`
- **`Query`**: `errorPolicy`
- **`Mutate`**: `errorPolicy`

All other options are ignored.
```

> Declaring those options ensures you provide values for them in your `defaultOptions`.

Reviewer: This answers my question about requiring this option to `ApolloClient`. Can you shuffle this a bit earlier and provide a code block showing the addition of `defaultOptions`? See my other comment.

> If you don't declare those options, TypeScript prevents you from using certain fields (for example, `errorPolicy` or `returnPartialData`) in the `ApolloClient` constructor's `defaultOptions`. However, you can still provide other default options.

Reviewer: I think it makes sense to move this into a `Note`? Might also be good to be a bit more specific on what to expect:

```suggestion
<Note>

If you don't provide type definitions for those options, their values are typed as `never` and TypeScript prevents you from assigning values to them in the `ApolloClient` constructor's `defaultOptions`. However, you can provide other default options.

</Note>
```

If you can't declare specific global defaults (for example, if you have multiple Apollo Client instances with different default options), you can opt out by declaring options as `unknown`:

```ts title="apollo.d.ts"
import type {} from "@apollo/client";

declare module "@apollo/client" {
  namespace ApolloClient {
    namespace DeclareDefaultOptions {
      interface WatchQuery {
        errorPolicy?: unknown;
        returnPartialData?: unknown;
      }
      interface Query {
        errorPolicy?: unknown;
      }
      interface Mutate {
        errorPolicy?: unknown;
      }
    }
  }
}
```

> Doing that allows you to call the `ApolloClient` constructor with any arbitrary `defaultOptions`.

Reviewer: ```suggestion
This allows you to call the `ApolloClient` constructor with arbitrary `defaultOptions`.
```

> In that case, all methods and hooks assume that those `defaultOptions` are set to the defaults, so you might get incorrect return values.

Reviewer: I'd wrap this in a caution block as this is a pretty big deal.

```suggestion
<Caution>

Use this sparingly because it reverts the behavior of any declared `defaultOptions` types to their defaults. This can result in incorrect runtime behavior that doesn't match the return type.

</Caution>
```
